src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -420,36 +418,47 @@ func isValidMacLogin(
 	return hmac.Equal(givenMac, expectedMAC), nil
 }
 
-// checkFlowsEqual checks if two registration flows have the same stages
-// within them. Order of stages does not matter.
-func checkFlowsEqual(aFlow, bFlow authtypes.Flow) bool {
-	a := aFlow.Stages
-	b := bFlow.Stages
-	if len(a) != len(b) {
-		return false
-	}
+// checkFlows checks a single flow a against another, b. If a contains at least
+// all of the stages that b does, checkFlows returns true.
+func checkFlows(a []authtypes.LoginType, b []authtypes.LoginType) bool {
+	// Sort the slices for simple comparison
 	sort.Slice(a, func(i, j int) bool { return a[i] < a[j] })
 	sort.Slice(b, func(i, j int) bool { return b[i] < b[j] })
 
-	// Account for any extra stages a user may do unnecessarily
-	extraStages := len(b) - len(a)
+	// Account for any extra stages a user may unnecessarily do.
+	extraStages := len(a) - len(b)
 	for i := range b {
```

Comment:
this doesn't look right.

Given:
```
a = { 'A', 'B' }
b = { 'Z' }
```

does this do the right thing (ie, return false)? Please add a test for it.

Also I think you're requiring that flows are completed in the same order as they are configured, which isn't necessary.

Reply:
Since they're sorted, the order in fact does not matter. You can play with the algo here:

https://play.golang.org/p/CHGHI03OBO

I've added the test case to both the playground and the .go file, both work :)

Reply:
```
	derivedFlows = []Flow{Flow{[]LoginType{LoginType("Z")}}}
	fmt.Println("Derived Flows:", derivedFlows)

	userFlow2 := Flow{[]LoginType{LoginType("A"), LoginType("B")}}
	fmt.Println("userFlow2:", userFlow2, checkFlowCompleted(userFlow2))
```

that gives true for me. I don't call that "working".

Reply:
Ahh, misread your original comment and switched the allowed and input flows. 

Yes, that input does result in the incorrect output!